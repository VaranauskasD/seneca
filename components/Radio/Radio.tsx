import React, { useState, Dispatch, SetStateAction, useEffect } from 'react'

import styled from 'styled-components'
// import { motion } from 'framer-motion'

interface RadioProps {
  id: number
  optionGroupId: number
  label: string
  name: string
  correct: boolean
  checked: boolean
  handleCheck: Dispatch<SetStateAction<number>>
  handleOption: (optionGroupId: number, correct: boolean) => void
  dynamicRatio: number
  groupQuantity: number
  isAnswered: boolean
}

const StyledRadio = styled.div<{
  $id: number
  $selected: boolean
  $dynamicRatio: number
  $groupQuantity: number
}>`
  min-width: 288px;
  padding: 20px 8px;
  color: ${(props) =>
    props.$selected
      ? `${props.theme.colors.secondary}`
      : `${props.theme.colors.primary}`};
  background: ${(props) =>
    props.$selected
      ? `rgb(
        ${
          props.theme.colors.dynamic.secondary.start.r +
          props.theme.colors.dynamic.secondary.range.r * props.$dynamicRatio
        }, 
        ${
          props.theme.colors.dynamic.secondary.start.g +
          props.theme.colors.dynamic.secondary.range.g * props.$dynamicRatio
        }, 
        ${
          props.theme.colors.dynamic.secondary.start.b +
          props.theme.colors.dynamic.secondary.range.b * props.$dynamicRatio
        })`
      : 'transparent'};
  border: 2px solid
    ${(props) =>
      props.$selected
        ? `rgb(
        ${
          props.theme.colors.dynamic.secondary.start.r +
          props.theme.colors.dynamic.secondary.range.r * props.$dynamicRatio
        }, 
        ${
          props.theme.colors.dynamic.secondary.start.g +
          props.theme.colors.dynamic.secondary.range.g * props.$dynamicRatio
        }, 
        ${
          props.theme.colors.dynamic.secondary.start.b +
          props.theme.colors.dynamic.secondary.range.b * props.$dynamicRatio
        })`
        : 'transparent'};
  border-radius: ${(props) => {
    if (props.$id === 0) {
      return `20px 20px 0px 0px`
    } else if (props.$id === props.$groupQuantity) {
      return `0px 0px 20px 20px`
    } else {
      return `100px`
    }
  }};

  @media (min-width: ${(props) => `${props.theme.breakpoints.sm}px`}) {
    border-radius: ${(props) => {
      if (props.$groupQuantity < 2) return `100px`
    }};

    min-width: 330px;
  }
  @media (min-width: ${(props) => `${props.theme.breakpoints.md}px`}) {
    min-width: 450px;
  }
`

// Animations To Be Completed

// const StyledRadioSelect = styled(motion.div)<{
//   $selected: boolean
//   $dynamicRatio: number
// }>`
//   position: relative;
//   width: 100%;
//   height: 100%;
// `

const StyledInput = styled.input`
  position: fixed;
  opacity: 0;
  pointer-events: none;
`

const StyledLabel = styled.label<{ isAnswered: boolean }>`
  ${(props) =>
    props.isAnswered
      ? `cursor: auto;`
      : `cursor: pointer;
        :hover {
        opacity: 0.8;
        }`}
`

export const Radio = (props: RadioProps) => {
  useEffect(() => {
    if (props.checked) {
      props.handleCheck(props.id)
      props.handleOption(props.optionGroupId, props.correct)
    }
  }, [])

  return (
    <React.Fragment>
      <StyledRadio
        $id={props.id}
        $selected={props.checked}
        $dynamicRatio={props.dynamicRatio}
        $groupQuantity={props.groupQuantity}
      >
        <StyledLabel isAnswered={props.isAnswered}>
          <StyledInput
            id={`option-${props.optionGroupId}-${props.id}`}
            type="radio"
            name={props.name}
            checked={props.checked}
            onChange={(event) => {
              props.handleCheck(props.id)
              props.handleOption(props.optionGroupId, props.correct)
            }}
            disabled={props.isAnswered}
          />
          {props.label}
        </StyledLabel>

        {/* <StyledRadioSelect /> */}
      </StyledRadio>
    </React.Fragment>
  )
}

export default Radio
