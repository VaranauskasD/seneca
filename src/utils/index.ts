export { theme } from './theme'
export { customRender } from './test'
