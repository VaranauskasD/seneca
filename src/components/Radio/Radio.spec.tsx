import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { customRender as render } from '../../utils/test'
import { Radio, RadioProps } from './Radio'

let mockRadioProps: RadioProps
beforeEach(() => {
  mockRadioProps = {
    id: 0,
    optionGroupId: 0,
    label: 'Cell wall',
    name: `option-0`,
    correct: false,
    checked: false,
    handleCheck: () => {},
    handleOption: () => {},
    dynamicRatio: 0,
    groupQuantity: 0,
    isAnswered: false,
  }
})

describe('Radio', () => {
  it('renders correctly', async () => {
    const { asFragment } = render(<Radio {...mockRadioProps} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
