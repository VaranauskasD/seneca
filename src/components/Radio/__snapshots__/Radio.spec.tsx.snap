// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ Sefou"
  >
    <label
      class="sc-dkzDqf hdgGtK"
    >
      <input
        class="sc-gsnTZi exMZOs"
        id="option-0-0"
        name="option-0"
        type="radio"
      />
      Cell wall
    </label>
  </div>
</DocumentFragment>
`;
